ceremonies/meetup-validation/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+/// This module is about finding which participants get their rewards based on their votes and attestations.
+/// The participant vecs are always vecs of participant ids
+/// The partitipant_vote and partcipant_attestations vecs (and their derived vecs) are indexed by the participant index
+/// ie. participant_votes[i] holds the vote of participant i
+
+pub fn get_updated_participants(
+	participants: &Vec<usize>,
+	participant_votes: &Vec<u32>,
+	participant_attestations: &Vec<Vec<usize>>,
+	attestation_threshold_fn: fn(usize) -> usize,
+) -> Result<UpdatedParticipants, MeetupValidationError> {
+	let mut updated_participants =
+		UpdatedParticipants { included: participants.clone(), excluded: vec![] };
+	updated_participants.exclude_participants(get_excluded_participants_no_vote(
+		&updated_participants.included,
+		participant_votes,
+	));
+
+	let (n_confirmed, _num_votes) =
+		find_majority_vote(&updated_participants.included, participant_votes)?;
+
+	updated_participants.exclude_participants(get_excluded_participants_wrong_vote(
+		&updated_participants.included,
+		participant_votes,
+		n_confirmed,
+	));
+
+	updated_participants.exclude_participants(get_excluded_participants_num_attestations(
+		&updated_participants.included,
+		participant_attestations,
+		attestation_threshold_fn,
+	));
+
+	Ok(updated_participants)
+}
+
+fn get_excluded_participants_no_vote(
+	participants: &Vec<usize>,
+	participant_votes: &Vec<u32>,
+) -> Vec<(usize, ExclusionReason)> {
+	let mut excluded_participants: Vec<usize> = vec![];
+	for i in participants {
+		match participant_votes[*i] {
+			v if v > 0 => continue,
+			_ => excluded_participants.push(*i),
+		}
+	}
+	excluded_participants
+		.into_iter()
+		.map(|p| (p, ExclusionReason::NoVote))
+		.collect()
```

Comment:
`collect` is O(n) so better try to avoid that:

```suggestion
	       let mut excluded_participants: Vec<(usize, ExclusionReason)> = vec![];
	       for i in participants {
		match participant_votes[*i] {
			v if v > 0 => continue,
			_ => excluded_participants.push((*i, ExclusionReason::NoVote)),
		}
	}
       excluded_participants
```